include/can.h

```diff
@@ -0,0 +1,380 @@
+/**
+ * @file
+ *
+ * @brief Public APIs for the CAN drivers.
+ */
+
+/*
+ * Copyright (c) 2018 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __DRIVERS_CAN_H
+#define __DRIVERS_CAN_H
+
+/**
+ * @brief CAN Interface
+ * @defgroup can_interface CAN Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/types.h>
+#include <device.h>
+#include <string.h>
+
+#define CAN_EX_ID      (1 << 31)
+#define CAN_MAX_STD_ID (0x7FF)
+#define CAN_STD_ID_MASK CAN_MAX_STD_ID
+#define CAN_EXT_ID_MASK (0x1FFFFFFF)
+#define CAN_MAX_DLC    (8)
+
+/* CAN_TX_* are the error flags from tx_callback and send.*/
+/** send successfully */
+#define CAN_TX_OK       (0)
+/** general send error */
+#define CAN_TX_ERR      (1)
+/** bus arbitration lost during sending */
+#define CAN_TX_ARB_LOST (2)
+/** controller is in bus off state */
+#define CAN_TX_BUS_OFF  (3)
+/** unexpected error */
+#define CAN_TX_UNKNOWN  (4)
+
+/** attach_* failed because there is no unused filter left*/
+#define CAN_NO_FREE_FILTER (-1)
+
+/** operation timed out*/
+#define CAN_TIMEOUT (1)
+
+/**
+ * @brief Statically define and initialize a can message queue.
+ *
+ * The message queue's ring buffer contains space for @a size messages.
+ *
+ * @param name Name of the message queue.
+ * @param size Number of can messages.
+ */
+#define CAN_DEFINE_MSGQ(name, size) K_MSGQ_DEFINE(name, sizeof(struct can_msg), size, 4)
+
+/**
+ * @typedef can_ide_t
+ * @brief Define if the message has a standard (11bit) or extended (29bit)
+ * identifier
+ */
+typedef enum {
+	CAN_STANDARD_IDENTIFIER,
+	CAN_EXTENDED_IDENTIFIER
+} can_ide_t;
+
+/**
+ * @typedef can_rtr_t
+ * @brief Define if the message is a data or remote frame
+ */
+typedef enum {
+	CAN_DATAFRAME,
+	CAN_REMOTEREQUEST
+} can_rtr_t;
+
+/**
+ * @typedef can_mode_t
+ * @brief Defines the mode of the can controller
+ */
+typedef enum {
+	/*Normal mode*/
+	CAN_NORMAL_MODE,
+	/*Controller is not allowed to send dominant bits*/
+	CAN_SILENT_MODE,
+	/*Controller is in loopback mode (receive own messages)*/
+	CAN_LOOPBACK_MODE,
+	/*Combination of loopback and silent*/
+	CAN_SILENT_LOOPBACK_MODE
+} can_mode_t;
```

Comment:
Please don't typedef these enumerations. This makes it unnecessarily confusing when using public APIs.

Instead, your ``can_configure(struct device *dev, can_mode_t mode, u32_t bitrate);`` would become ``can_configure(struct device *dev, enum can_mode mode, u32_t bitrate);`` which makes it painfully obvious to anyone writing / looking at a driver what the value type is being passed around.

See: 
* http://docs.zephyrproject.org/contribute/contribute_guidelines.html#coding-style
* https://www.kernel.org/doc/html/latest/process/coding-style.html#typedefs

Reply:
Thanks for the hint. I changed that.